c7n/resources/s3.py

```diff
@@ -3475,3 +3475,122 @@ def process_bucket(self, b):
                 raise
             controls = {}
         b[self.annotation_key] = controls.get('OwnershipControls')
+
+
+@filters.register('bucket-replication')
+class BucketReplication(Filter):
+    """Filters for S3 buckets that have replication-rule
+
+    Reference: https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication.html
+
+    :example
+
+    Find buckets with replication rule enabled
+
+    .. code-block:: yaml
+
+            policies:
+              - name: s3-bucket-replication-cross-region
+                resource: s3
+                region: us-east-1
+                filters:
+                  - type: bucket-replication
+                    state: True
+                    destination: cross-region
+              - name: s3-bucket-replication-same-region
+                resource: s3
+                region: us-east-1
+                filters:
+                  - type: bucket-replication
+                    state: True
+                    destination: same-region
+              - name: s3-bucket-has-replication
+                resource: s3
+                region: us-east-1
+                filters:
+                  - type: bucket-replication
+                    state: True
+    """
+    schema = type_schema('bucket-replication',
+                         state={'type': 'boolean'},
+                         destination={'type': 'string', 'enum': ['cross-region', 'same-region']})
+
+    permissions = ("s3:GetReplicationConfiguration",)
+    annotation_key = 'c7n:bucket-replication'
```

Comment:
We already pull replication configuration during [`assemble_bucket()`](https://github.com/cloud-custodian/cloud-custodian/blob/d35f529764af6435e8848f0ed5b4b9484779ef1a/c7n/resources/s3.py#L420-L488) because we include location in the base s3 [augment table](https://github.com/cloud-custodian/cloud-custodian/blob/d35f529764af6435e8848f0ed5b4b9484779ef1a/c7n/resources/s3.py#L403).

I think that means we _could_ use `Location` rather than `c7n:bucket-replication` for the annotation key here, though your choice is more of what we'd expect to see elsewhere. Maybe copying `Location` --> `c7n:bucket-replication` as part of this filter's evaluation would feel more natural but also avoid redundant calls :thinking:

Reply:
Thanks for the comment. PR is updated.